frame/support/src/weights/weight_v2.rs

```diff
@@ -0,0 +1,451 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use core::ops::{Add, AddAssign, Div, Mul, Sub, SubAssign};
+use sp_runtime::{
+	traits::{Bounded, CheckedAdd, CheckedSub, One, Zero},
+	RuntimeDebug,
+};
+
+use super::*;
+
+#[derive(
+	Encode, Decode, MaxEncodedLen, TypeInfo, Eq, PartialEq, Copy, Clone, RuntimeDebug, Default,
+)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct Weight {
+	/// The weight of computational time used.
+	computation: ComputationWeight,
+	/// The weight of the storage bandwidth used.
+	bandwidth: BandwidthWeight,
+}
+
+impl Weight {
+	/// Create a new Weight with zero computation and bandwidth.
+	pub const fn new() -> Self {
+		Self { computation: 0, bandwidth: 0 }
+	}
+
+	/// Set the computational part of the weight.
+	pub const fn set_computation(mut self, c: ComputationWeight) -> Self {
+		self.computation = c;
+		self
+	}
+
+	/// Set the bandwidth part of the weight.
+	pub const fn set_bandwidth(mut self, b: BandwidthWeight) -> Self {
+		self.bandwidth = b;
+		self
+	}
+
+	/// Return the computational part of the weight.
+	pub const fn computation(&self) -> ComputationWeight {
+		self.computation
+	}
+
+	/// Return the bandwidth part of the weight.
+	pub const fn bandwidth(&self) -> BandwidthWeight {
+		self.bandwidth
+	}
+
+	/// Return a mutable computational part of the weight.
+	pub fn computation_mut(&mut self) -> &mut ComputationWeight {
+		&mut self.computation
+	}
+
+	/// Return a mutable bandwidth part of the weight.
+	pub fn bandwidth_mut(&mut self) -> &mut BandwidthWeight {
+		&mut self.bandwidth
+	}
+
+	pub const MAX: Self =
+		Self { computation: ComputationWeight::MAX, bandwidth: BandwidthWeight::MAX };
+
+	/// Get the conservative min of `self` and `other` weight.
+	pub fn min(&self, other: Self) -> Self {
+		Self {
+			computation: self.computation.min(other.computation),
+			bandwidth: self.bandwidth.min(other.bandwidth),
+		}
+	}
+
+	/// Get the aggressive max of `self` and `other` weight.
+	pub fn max(&self, other: Self) -> Self {
+		Self {
+			computation: self.computation.max(other.computation),
+			bandwidth: self.bandwidth.max(other.bandwidth),
+		}
+	}
+
+	/// Try to add some `other` weight while upholding the `limit`.
```

Comment:
Needs explicitly mentioning what happens in case of overflow or exceeding any of the limits